Add property-based tests to ensure C and Python implementations of the same algorithms behave identically

nchammas opened the pull request:
This PR adds a suite of property-based tests that ensure that the C and Python implementations of the same algorithms behave identically. 

The tests show that we have a lot of cleanup work to do, with Hypothesis uncovering failing test cases for all but 3 of Jellyfish's algorithms:

```
$ pytest -v jellyfish/test_properties.py 
====================================================== test session starts =======================================================
platform darwin -- Python 3.5.2, pytest-3.0.4, py-1.4.31, pluggy-0.4.0 -- python3
cachedir: .cache
rootdir: .../jellyfish, inifile: 
plugins: hypothesis-3.6.1
collected 11 items 

jellyfish/test_properties.py::test_py_soundex_equals_c_soundex FAILED
jellyfish/test_properties.py::test_py_nysiis_equals_c_nysiis FAILED
jellyfish/test_properties.py::test_py_match_rating_codex_equals_c_match_rating_codex FAILED
jellyfish/test_properties.py::test_py_metaphone_equals_c_metaphone FAILED
jellyfish/test_properties.py::test_py_porter_stem_equals_c_porter_stem FAILED
jellyfish/test_properties.py::test_py_levenshtein_distance_equals_c_levenshtein_distance PASSED
jellyfish/test_properties.py::test_py_damerau_levenshtein_distance_equals_c_damerau_levenshtein_distance FAILED
jellyfish/test_properties.py::test_py_hamming_distance_equals_c_hamming_distance PASSED
jellyfish/test_properties.py::test_py_jaro_distance_equals_c_jaro_distance FAILED
jellyfish/test_properties.py::test_py_jaro_winkler_equals_c_jaro_winkler PASSED
jellyfish/test_properties.py::test_py_match_rating_comparison_equals_c_match_rating_comparison Segmentation fault: 11
```

The first problem to be addressed, in my view, is the segmentation fault reported in #73, since that interrupts the test run and prevents Hypothesis from providing a clean report of failing test cases.

Here are some of the failing test cases:

```
$ pytest -s jellyfish/test_properties.py 
====================================================== test session starts =======================================================
platform darwin -- Python 3.5.2, pytest-3.0.4, py-1.4.31, pluggy-0.4.0
rootdir: .../jellyfish, inifile: 
plugins: hypothesis-3.6.1
collected 11 items 

jellyfish/test_properties.py
Falsifying example: test_py_soundex_equals_c_soundex(s='ı')
Falsifying example: test_py_nysiis_equals_c_nysiis(s='\x80')
Falsifying example: test_py_match_rating_codex_equals_c_match_rating_codex(s='\x80')
Falsifying example: test_py_metaphone_equals_c_metaphone(s='0H')
Falsifying example: test_py_porter_stem_equals_c_porter_stem(s='\x00')
Falsifying example: test_py_damerau_levenshtein_distance_equals_c_damerau_levenshtein_distance(s1='Ā', s2='')
Falsifying example: test_py_jaro_distance_equals_c_jaro_distance(s1='100200', s2='021')
```

This PR updates the Travis CI build to run these tests, so if we merge this PR in that will technically break the build. We can do that, or we can keep this PR open and periodically re-run the build to track our progress as we clean up these bugs, merging it in only when everything has been taken care of.

Fixes #68.

nchammas commented on tox.ini:
```diff
@@ -1,47 +1,53 @@
+deps =
+    pytest
+    unicodecsv
+    hypothesis>=3.6.1,<4.0.0
```
Tox lets you specify dependencies once and then refer to them as a group, so I've refactored accordingly.

nchammas commented on tox.ini:
```diff
@@ -1,47 +1,53 @@
-    py.test jellyfish/test.py
-deps = pytest
-    unicodecsv
+    pytest jellyfish/test.py jellyfish/test_properties.py
```
pytest updated its invocation method to drop the confusing `.` so I've updated this accordingly.

nchammas commented on tox.ini:
```diff
@@ -1,47 +1,53 @@
+
 [testenv:flake8]
-basepython=python2.7
+basepython = python2.7
```
I've added whitespace around the `=` to make it consistent with the rest of the file.

nchammas commented:
@jamesturk - Let me know how you would like to proceed.

As noted earlier, the build is failing due to bugs that have been uncovered. You won't be able to see the falsifying tests cleanly in Travis's output until the seg fault bug has been fixed, since that interrupts Tox. But you can always run pytest independently with `-vs` to get that output.

nchammas commented on jellyfish/test_properties.py:
```diff
@@ -0,0 +1,83 @@
+
+
+@given(s1=text(), s2=text())
+@settings(verbosity=Verbosity.verbose)
```
We'll want to remove this line once we fix the seg fault in `match_rating_comparison()`.

nchammas commented:
@jamesturk - Any initial reaction to this PR or rough plan to tackle the issues uncovered by it?

jamesturk commented:
I definitely think it is a good idea, but I think it'll need to remain in a separate branch until we're able to tackle the failing tests.  I haven't had much free time to work on this lately, but I'd be glad to accept patches, or I'll try to get to the edge cases as soon as I can.

nchammas commented:
Trying to trigger a new Travis CI build. Will close/reopen this PR.

nchammas commented:
Hmm, looks like Travis is not running the new tests from this PR anymore. Looking at the commit history, it looks like this commit is the culprit: 60b2d88ba6e3d7f1b464f514d0163814ce5411a6

@jamesturk - Any reason we don't want to delegate the test running to Tox? Are we not using the tox.ini anymore?

jamesturk commented:
Running tests on travis from Tox had a few disadvantages in build time and reporting and some weird spurious failures, but I don't think that was affecting it running tests on PRs.

Looks like Travis has run it though

nchammas commented:
Nope, Travis is not running the tests introduced in this PR because .travis.yml doesn't mention them, only tox.ini does. This raises a minor issue, which is that we now have to specify what tests to run in 2 separate places. This hasn't been a problem to date because there is only one test file in master, but it's exposed here since I'm adding a new test file.

Not a big deal. I can either add the new file explicitly to the Travis file, or we can tweak how pytest is invoked so that it automatically picks up new test files as they are added.

jamesturk commented:
Ah sorry, missed that- mind adding it to Travis?  I figure that's easiest

On Fri, Jan 5, 2018 at 4:30 PM, Nicholas Chammas <notifications@github.com>
wrote:

> Nope, Travis is not running the tests introduced in this PR because
> .travis.yml doesn't mention them, only tox.ini does. This raises a minor
> issue, which is that we now have to specify what tests to run in 2 separate
> places. This hasn't been a problem to date because there is only one test
> file in master, but it's exposed here since I'm adding a new test file.
>
> Not a big deal. I can either add the new file explicitly to the Travis
> file, or we can tweak how pytest is invoked so that it automatically picks
> up new test files as they are added.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/jamesturk/jellyfish/pull/76#issuecomment-355670463>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAAfYmcCVUKAKJBIqw3JBHPw6fPEY7XDks5tHpRzgaJpZM4LVDuG>
> .
>

nchammas commented:
I cleaned up how we run tests according to what I got out of your comments on this PR.
* I removed tox.ini entirely.
* I captured our test dependencies in a dedicated file and referenced that from .travis.yml.
* I updated how we invoke pytest so that `test_*` files are automatically picked up.
* I updated the README to reference pytest and not tox.

I now see that the Hypothesis tests are being run by Travis as expected. I also see that we were [failing some flake8 tests](https://travis-ci.org/jamesturk/jellyfish/jobs/351101543#L500-L501). I added directives to ignore those two specific errors since they seem to be expected.

@jamesturk - This PR has been open for over a year. I'd like to get it merged in or close it. Since it seems the problems revealed by these new tests won't be fixed soon, I think the only way to get this PR merged in is if we disable all the new tests. I'm fine with that. As you or contributors fix problems, they can progressively enable these new tests.

Does that work for you? I'm looking for specific requirements I can meet to get this PR merged in. If the best approach is to close it without merging it in, I'm fine with that too. Not every contribution works out, after all.

One way or the other, I'd like a decisive answer.
